### DISCUSSION TITLE
Incentivized, ordered, batched and cheap execution of internal vouchers.

### DISCUSSION THREAD

**Discussion by GCdePaula** [cartesi/cips#28] [Ideas]
Here are some ideas that have been on my mind lately. I’m not sure whether they make sense, whether they work at all, or whether it’s an improvement. I’ll attempt to outline them here so we can get a discussion going. I'm not convinced they are good ideas, and recent conversations with @augustoteixeira suggested that we may have other solutions which achieve the same things. But documenting discussions are a positive endeavor in any case.

Thanks to @felipeargento and @augustoteixeira for previous discussions on this topic!

## Introduction

Currently, we treat every voucher the same way. They are a pair of target address and payload, which are arbitrary values. Vouchers are executed from the context of the rollups contract, where the `msg.sender` is the rollups contract address. This allows for permissioned calls to the portal that contains all the rollups' locked assets.

Concretely, once an epoch is finalized, vouchers can be executed in any order. The finalized epoch hash contains a Merkle tree, which allows for random access of vouchers. Additionally, we have in storage the set of vouchers which have been executed, so that one cannot execute the same voucher multiple times.

The code that executes outputs inside the rollups doesn’t understand the contents of vouchers, it simply dispatches them. When the dApp wants to withdraw money to some account, it emits a voucher where the target address is the portal. This will dispatch a message call to the portal, which will check whether the sender is the rollups, reverting otherwise. The payload of the voucher contains the selector and arguments that will make the portal transfer the assets. In the case of ERC20 transfers, it will trigger a call to yet another contract.

## Issues with current design

There are two separate characteristics of voucher execution I wish to address with this discussion.

* No ordering guarantees. This has all sorts of sweeping consequences for the entire rollups. It can be seen as a feature rather than a bug, since it solves a lot of issues created if ordering is imposed. However, we lose some guarantees on the entire system, related to being sure there will be enough money in the portal during a withdrawal.

    One consequence is related to bank runs. If the dApp generates more withdrawals than the portal has (itself a bug or a malicious behavior), some vouchers won’t go through. Imagine a liquidity provider who’s providing liquidity to some dApp. The issue is that if they observe a voucher transferring money to the liquidity smart contract, they can’t be sure this voucher will go through. The reason it may not go through is because the dApp may emit a voucher transferring all the assets from the portal later on the epoch, and that voucher can be executed before the provider’s voucher since there’s no ordering. This imposes the necessity of liquidity providers trusting the dApp. With ordering, this trust is not necessary, since vouchers emitted later can only be executed later.

* Costs. There are seven costs associated with executing a voucher that we can address, which I tried to list in growing order of gas cost. I haven’t measured any of this, so the ordering may be wildly inaccurate.
    1. Voucher calldata. This is varies a lot, but let us consider only simple stuff like ERC20 token transfer. It includes the target address (which is the portal), and the function selector of the portal (say, "withdrawToken"). This cost is low, but we can reduce it further.
    2. External function call to a different contract. This cost is low, but we can reduce it further.
    3. Verify Merkle proof.
    4. Storage read to check whether the voucher has been executed.
    5. Storage write to mark the voucher as executed. 
    6. Base cost of transaction.
    7. Merkle proof calldata cost. If my math is correct, we need to supply 48 siblings (totaling 1536 bytes of calldata) to prove a single voucher, which is quite expensive.

It should be clear that it is incredibly expensive to execute a voucher. There are mitigating solutions, mainly related to batching. However, batching is restricted by the large 1536 bytes of call data necessary even for simple token transfers; the size of our batches will be limited by this. Additionally, not all items in that list will be reduced by batching with the current design. There is perhaps a way to share siblings of Merkle proofs between batched inputs, but they are not clear in my head, and I’m unsure how much this will reduce costs.

## Proposal

There’s a subset of vouchers that perhaps will compromise the majority of executed vouchers, where the target address is the portal (in the diamond design, it’s the rollups address itself). These are what we’ve been calling internal vouchers, and consist mainly of asset transfers. Their scope is known and limited, and they are generally well behaved.

The proposal starts with splitting vouchers from internal vouchers, and significantly changing how the latter works. This means there would still be external vouchers, which work as our current vouchers for the general case.

Starting with voucher execution involving the portal. Instead of the rollups issuing a function call to itself, with an arbitrary payload that will concretely execute something well defined, we change the rollups to actually understand vouchers, and execute a known set of functionalities. That is, the internal voucher’s format becomes something that the rollups/portal understands, consisting of a selector (which has a known range of values, describing what we want to do, such as an ERC20 token transfer), and arguments (which, once we know the selector, has a known type and well defined values).

The analogy in my head is switching from calling `eval` on an arbitrary string in the same execution context as the host program (essentially what we’re doing, except perhaps adding the assumption that whoever generated that arbitrary string can be trusted), to reading a well defined input and executing a known set of functionalities.

(A note on removing access to the portal from external vouchers. If internal vouchers are the only thing that requires access to the portal, we can remove this access from external vouchers, possibly increasing the security of the rollups.)

(A note on gas usage and external calls. We still need to make external calls when executing transfers like ERC20 token transfer. Since the target contract can be malicious, we must forward a constant, restricted amount of gas. If the gas is not enough, the voucher is still considered as executed. Also, since the forwarded gas is a constant, a malicious executor may not try to "force" a voucher to be skipped.)

---

Then, we impose an ordering to vouchers; they must be executed in the order they were emitted. This solves the bank run issue, which is quite important. (Although it generates a lot of issues in itself). It also removes the necessity for random access; we no longer need Merkle trees, we can use a rolling hash, which greatly reduces gas costs.

(A note on failed voucher execution. If a voucher failed to execute for whatever reason, it must be considered executed, since reverting can mean halting all further voucher executions because of ordering.)

The proposed rolling hash structure is a back-to-front rolling hash of all vouchers, and is calculated thus:

```
H[0] = bytes32(0x0)
H[1] = keccak(H[0] .. keccak(V[n]))
H[2] = keccak(H[1] .. keccak(V[n-1]))
…
H[n] = keccak(H[n-1] .. keccak(V[0]))
```

This changes the structure of the rollups epoch hash. Besides the machine state root Merkle hash and the outputs root Merkle hash (for external vouchers), the epoch hash would have to include this rolling hash.

(A note on the verification game. Since this rolling hash is back to front, it has to be recalculated from scratch after every input, which doesn’t bode well for arbitration. However, we can solve this through partitioning, similarly to how we find which instruction was dishonest.)

The setup for execution vouchers starts by figuring out the `H[n]` from the epoch hash. We must have in storage the current rolling hash, which starts with the value `H[n]`.

Here’s a short description of a pseudo-algorithm for executing an internal voucher. I’d describe it as successively popping vouchers from the voucher stack, where the latter is stored in a rolling hash structure.
1. The user sends to the blockchain the content of the voucher at the top of this pseudo-stack, as well as the next rolling hash. No need to send 1536 bytes of calldata for Merkle proof.
2. The verification process is hashing the voucher and the next rolling hash together (as described above), and comparing it with the current rolling hash we have in storage. If they match, all’s good, otherwise revert. This should be a lot cheaper than validating a Merkle proof as we’re currently doing.
3. Execute the voucher the way described before, without making the extra external function call.  If the execution fails for whatever reason, don’t revert and consider the voucher executed. This behavior is analogous to reverted transactions in Ethereum, which are still considered mined and added to the chain.
4. Update storage of current rolling hash to the value of next rolling hash.

This should greatly reduce the costs of executing vouchers.

(A note on reentrancy. We may still need to keep the reentrancy guard in place. Since ERC20 token transfer may execute malicious code, it could conceivably reenter, which could conceivably be an attack. It’s possible that swapping steps 3 and 4 may solve this issue, but it removes an important optimization when batching.)

---

Now let’s do batching. The setup is the same, but the execution now has a loop.
1. The user sends to the blockchain an array of pairs `(voucher, next rolling hash)` in order, instead of a single pair.
2. Load the current rolling hash from storage to memory.
3. In a loop, iterating over the array, run these steps:
    1. Verify the next voucher, as described before. If the proof is wrong at any iteration, we can either revert everything or just break from the loop.
    2. Update in memory the current rolling hash.
    3. Execute the voucher. If the execution fails for whatever reason, keep moving.
4. Update storage of current rolling hash to what’s on memory by the end of the loop.

Comparing with the previously written list of the current costs:
1. Since we are not doing the message call dispatch, we can use a smarter encoding which does not include the address (we know it’s internal) and saves on the selector (it has a known list of value). This saving is quite marginal, to be fair.
2. No longer executing extra external function call.
3. Proof verification is much cheaper.
4. In batched execution, we shift from reading storage at every voucher to reading storage only once per batch.
5. In batched execution, we shift from writing storage at every voucher to writing storage only once per batch.
6. In batched execution, base cost of transaction per output is greatly reduced. Although this is also true for batched executions in the current design, batches can be larger because we are reducing the block space usage due to all other items. In comparison, this further dilutes/amortizes the base transaction cost per output.
7. Proof is now just a single hash, instead of 48 hashes.

This change gets us ordering and saves on costs. What’s the rub?

## Issues with proposal

Well, ordering is the rub. It has benefits, but there are reasons why we chose unordered vouchers. First, if one voucher gets stuck, we halt the entire rollups. Second, if you’re interested in voucher number one thousand, you’d need to execute nine-hundred and ninety-nine vouchers to get to yours. Quite appalling. But not all is lost.

On the first issue, this proposal is not for all vouchers; it’s the subset of vouchers on asset transfer that are handled internally. This makes the vouchers much well behaved, in the sense that we can make them never get stuck. The idea is that if the voucher reverts, we may mark it as executed without mercy and keep going on. We’d still have the generic vouchers as they are currently, if needed.

The second issue is more delicate. The solution to this is some sort of automatic voucher execution system, which we already wanted to create anyway. But here it’s more necessary and more critical. Out of order vouchers may always be altruistically executed; it’s a harder sell when altruistically executing a voucher requires executing the preceding nine-hundred and ninety-nine vouchers.

There are three issues related to automatic execution:

* Unpredictable and manipulatable voucher execution behavior. Since currently voucher’s value is arbitrary, consisting of the target address and payload (which includes the function selector and arguments), really anything can happen. The gas cost of the execution can really be anything, and predicting it is largely unreliable. This makes it harder to create an incentive system around voucher execution.
* Race condition. To create an automated voucher execution system, we incentivize executors to execute vouchers in some way. Adding this incentive generates a race condition, because only the first transaction will be rewarded and all the others will revert.
* Gas price spikes, or other fluctuations that may make the incentive not enough.

The first issue isn’t an issue at all with internal vouchers. Gas costs are well bounded and well behaved.

The second issue I thought I had a solution, but it doesn’t work. I’ll outline it here anyway because I think it’s interesting and may serve us in the future. A permissionless incentivized call that rewards only the first transaction has a race condition; the Nash equilibrium is weird because everyone will want to do it, but because of this no one will, and so on. However, a permissionless incentivized call that rewards the first N transactions may have a better behaved equilibrium. The idea was to make batched voucher execution run whatever was in the top of the pseudo-stack. If someone front runs me, I’d still get my money because I’d be executing the next batch. The issue is that the execution transaction contains the proofs, and the proofs change when the pseudo-stack changes, so this is no use, the race condition is still there. We'll have to use another solution for this, but we are already addressing this issue for the current design so I'm sure we have something.

The third issue I have no idea how to deal with. I don’t even know how bad it is; maybe it’s not really an issue and just considerably over-paying executors will be enough.

I propose a system for the automatic execution of batched internal vouchers, which is delegating the incentive to the dApp developer. Imagine our batched execution has a fixed size of 128 vouchers. The solution is for the dApp developer for every 127 outputs emit an additional voucher, which transfers money from the rollups portal to the voucher executor. The money to pay for the execution comes from the withdrawal input; the dApp developer has to charge a fee from users whenever they want to withdraw, to pay for the automatic execution. I don’t know how to estimate these fees and deal with gas price spikes. I’m probably missing some details with this solution.

## Conclusion

This requires a lot of consideration. The details are fuzzy, and the consequences sweeping. Nevertheless, I believe there may be some merit in this idea and that we should think about it, even if it’s just to throw it away and keep just our current design.

**Comment by guidanoli** [comment]
> Also, since the forwarded gas is a constant, a malicious executor may not try to "force" a voucher to be skipped.

Couldn't this break with changes to gas prices ([SWC-134](https://swcregistry.io/docs/SWC-134))?
